docs(webpack): Add docs for webpack ci deployment (#21650)
nrwl/nx: docs/shared/recipes/ci-deployment.md

- `@nx/js:tsc`
- `@nx/next:build`

{% /tab %}
{% /tabs %}

## Programmtic usage

> If you using a custom setup that does not support the creation of a `package.json` or a lock file, you can still use Nx to generate them via The `createPackageJson` and `createLockFile` functions which are exported from `@nx/js`:

Reviewer: ```suggestion
If you are using a custom setup that does not support the creation of a `package.json` or a lock file, you can still use Nx to generate them via The `createPackageJson` and `createLockFile` functions which are exported from `@nx/js`:
```

} = require('@nx/devkit');
const {
  createLockFile,
  createPackageJson,
  getLockFileName,
} = require('@nx/js');
const { writeFileSync } = require('fs');

async function main() {
  const outputDir = 'dist'; // You can replace this with the output directory you want to use
  // Detect the package manager you are using (npm, yarn, pnpm)
  const pm = detectPackageManager();
  let projectGraph = readCachedProjectGraph();
  if (!projectGraph) {
    projectGraph = await createProjectGraphAsync();
  }
  // You can replace <NX_TASK_TARGET_PROJECT> with the name of the project if you want.
  const projectName = process.env.NX_TASK_TARGET_PROJECT;
  const packageJson = createPackageJson(projectName, projectGraph, {
    isProduction: true, // Used to strip any non-prod dependencies
    root: projectGraph.nodes[projectName].data.root,
  });

  const lockFile = createLockFile(
    packageJson,
    projectGraph,
    detectPackageManager()
  );

  const lockFileName = getLockFileName(pm);

  writeFileSync(`${outputDir}/package.json`, packageJson);
  writeFileSync(`${outputDir}/${lockFileName}`, lockFile, {
    encoding: 'utf8',
  });

  //... Any additional build steps you want to run

>   exec(`${pm.exec} nx build`, (err, stdout, stderr) => {

Reviewer 2: I don't think we need call the `nx build` here, users can do `nx build && npm run custom-script`. Note that most `build` commands will also clear dist first, so copy definitely needs to be second.

Author: I wanted to show that you could run it solely from script. Since `nx build && nx copy-packagejson` is usually common knowledge.

    if (err) {
      console.error(err);
      return;
    }
    console.log(stdout);
  });
}

main();
```

Then to run the script, update your `package.json` to include the following:

```json {% fileName="package.json" %}
{
  "scripts": {

>     "build": "node scripts/custom-build.js"

Reviewer 2: Maybe this should be `custom-build` rather than `build` to differentiate between build targets that could be inferred or configured in `project.json`?

Reviewer 2: Or maybe just `copy-package-json` script name?

Author: it is `custom-build`

  // do any additional manipulations to "package.json" here

  const lockFile = createLockFile(packageJson);
  writeJsonFile(`${options.outputPath}/package.json`, packageJson);
  writeFileSync(`${options.outputPath}/${packageLockFileName}`, lockFile, {
    encoding: 'utf-8',
  });

  // any subsequent executor code
}
```

{% /tab %}
{% /tabs %}

> {% callout type="note" title="What about Vite ?" %}

Reviewer: ```suggestion
{% callout type="note" title="What about Vite?" %}
```

> Vite is a build tool that is great for development, and we want to make sure that it is also great for production. We are working on a plugin for Vite that will have parity with the Webpack plugin. Stay tuned for updates.

Reviewer: ```suggestion
Vite is a build tool that is great for development, and we want to make sure that it is also great for production. We are working on an `NxVitePlugin` plugin for Vite that will have parity with the `NxWebpackPlugin`. Stay tuned for updates.
```
